### PR TITLE
Port unblock user action #2471647

### PR DESCRIPTION
**Pull request by lu4nik**
Drupal.org issue: https://www.drupal.org/node/2471647

### CODE REVIEW

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  const BLOCKED = FALSE;
+
+  /**
+   * The action to be tested.
+   *
+   * @var \Drupal\rules\Engine\RulesActionInterface
+   */
+  protected $action;
+
+
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php:53]: unnecessary empty line

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  /**
+   * @dataProvider userProvider
+   * @covers ::execute
+   */
+  public function testUnblockUser($active, $authenticated, $expects) {
+    $user = $this->getMock('Drupal\user\UserInterface');
+
+    $user = $this->getMock('Drupal\user\UserInterface');
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php]: duplicated

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  /**
+   * The action to be tested.
+   *
+   * @var \Drupal\rules\Engine\RulesActionInterface
+   */
+  protected $action;
+
+
+  /**
+   * @dataProvider userProvider
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php]: a small comment about the function would be nice

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  public function testUnblockUser($active, $authenticated, $expects) {
+    $user = $this->getMock('Drupal\user\UserInterface');
+
+    $user = $this->getMock('Drupal\user\UserInterface');
+
+    $user->expects($this->any())
+      ->method('isBlocked')
+      ->willReturn(!$active);
+
+
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php]: empty line

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  /**
+   * The action to be tested.
+   *
+   * @var \Drupal\rules\Engine\RulesActionInterface
+   */
+  protected $action;
+
+
+  /**
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php:54]: a small comment about the function would be nice

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+   * @dataProvider userProvider
+   * @covers ::execute
+   */
+  public function testUnblockUser($active, $authenticated, $expects) {
+    $user = $this->getMock('Drupal\user\UserInterface');
+
+    $user = $this->getMock('Drupal\user\UserInterface');
+
+    $user->expects($this->any())
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php]: once makes more sense here

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      ->method('isBlocked')
+      ->willReturn(!$active);
+
+
+    $user->expects($this->any())
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php]: once makes more sense here as well. any means 0 or more. let's be a bit more strict

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\Tests\rules\Integration\Action\UserUnBlockTest.
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php]: should be UserUnblockTest

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+ */
+
+namespace Drupal\Tests\rules\Integration\Action;
+
+use Drupal\Tests\rules\Integration\RulesEntityIntegrationTestBase;
+
+/**
+ * @coversDefaultClass \Drupal\rules\Plugin\Action\UserUnBlock
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php]: should be UserUnblock

File: src/Plugin/Action/UserUnblock.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\Plugin\Action\UserUnBlock.
```
Comment (milkovsky) [inline src/Plugin/Action/UserUnblock.php]: should be UserUnblock

File: src/Plugin/Action/UserUnblock.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+ *     "user" = @ContextDefinition("entity:user",
+ *       label = @Translation("User"),
+ *       description = @Translation("Specifies the user, that should be unblocked.")
+ *     )
+ *   }
+ * )
+ *
+ * @todo: Add access callback information from Drupal 7.
+ */
+class UserUnBlock extends RulesActionBase {
```
Comment (milkovsky) [inline src/Plugin/Action/UserUnblock.php]: again UserUnblock

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+ * @group rules_actions
+ */
+class UserUnblockTest extends RulesEntityIntegrationTestBase {
```
Comment (milkovsky) [inline tests/src/Integration/Action/UserUnblockTest.php:16]: UserUnblockTest

**Comment by milkovsky** [comment]
looks good for me

File: src/Plugin/Action/UserUnblock.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    public function summary() {
+        return $this->t('Unblock a user');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function execute() {
+        /**
+         * @var $user \Drupal\user\UserInterface
```
Comment (klausi) [inline src/Plugin/Action/UserUnblock.php]: We are using the one line syntax for such annotations within function bodies, i.e. /*\* @var $user \Drupal\user\UserInterface */

File: src/Plugin/Action/UserUnblock.php [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+         */
+        $user = $this->getContextValue('user');
+
+        // Do nothing if user is anonymous or isn't blocked.
+        if ($user->isAuthenticated() && $user->isBlocked()) {
+            $user->activate();
```
Comment (klausi) [inline src/Plugin/Action/UserUnblock.php]: indentation error, use 2 spaces.

Comment (klausi) [inline src/Plugin/Action/UserUnblock.php]: Also you should mark the user for auto saving now, see the UserRoleAdd action for example.

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+      ->method('isAuthenticated')
+      ->willReturn($authenticated);
+
+    $user->expects($this->{$expects}())
+      ->method('activate');
+
+    $this->action->setContextValue('user', $user);
+
+    $this->action->execute();
+  }
```
Comment (klausi) [inline tests/src/Integration/Action/UserUnblockTest.php:92]: After that we should call autoSaveContext() on the action and make sure it returns the correct user context key or not.

File: tests/src/Integration/Action/UserUnblockTest.php [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      [self::ACTIVE, self::AUTHENTICATED, 'never'],
+      // Test blocked anonymous user.
+      [self::BLOCKED, self::ANONYMOUS, 'never'],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp() {
```
Comment (klausi) [inline tests/src/Integration/Action/UserUnblockTest.php]: The setup method should be the first one to be consistent with other tests.

**Comment by lu4nik** [comment]
Hi @klausi 
I've fixed above problems, but not sure about the autoSaveContext check. Maybe it is better to write 4 separate test methods instead one with @dataProvider, what do you think? 

Also I suppose that userBlock action should be updated accordingly (use auto save).

**Comment by klausi** [comment]
Merged, thanks!

**Pull request closed without merging**